agent/engine/dockerclient/dockerclientfactory.go

```diff
@@ -121,8 +135,66 @@ func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
 		err = client.Ping()
 		if err != nil {
 			log.Infof("Failed to ping with Docker version %s: %v", version, err)
+			continue
 		}
 		clients[version] = client
 	}
 	return clients
 }
+
+func findDockerVersionsfromMinMaxVersions(endpoint string) (map[DockerVersion]dockeriface.Client, error) {
+	// get a Docker client with the default supported version
+	client, err := newVersionedClient(endpoint, string(minDockerAPIVersion))
+	if err != nil {
+		return nil, errors.Wrap(err, "min API version check: error while creating default Docker client")
+	}
+
+	clientVersion, err := client.Version()
+	if err != nil {
+		return nil, errors.Wrap(err, "min API version check: error while getting client version")
+	}
+
+	clients := make(map[DockerVersion]dockeriface.Client)
+	// check if the docker.Env obj has MinAPIVersion key
+	if clientVersion.Exists(minAPIVersionKey) {
+		minAPIVersion := clientVersion.Get(minAPIVersionKey)
+		apiVersion := clientVersion.Get(apiVersionKey)
+		for _, version := range getKnownAPIVersions() {
+			dockerClient, err := getDockerClientForVersion(endpoint, string(version), minAPIVersion, apiVersion)
+			if err != nil {
+				log.Debugf("Unable to get Docker client for version %s: %v", version, err)
```

Comment:
I think this should be `Infof`

Reply:
It makes more sense to have it as `Debugf` to me, since we are just trying to obtain docker client for a particular version and it might not always be successful, which is expected. All the older versions' agent logs would contain these messages looking like errors in the beginning. So made it `Debugf`. What do you think?

Reply:
If there are asynchronous failures for containers because of a specific versioned docker client  not being there, having these errors printed out during discovery helps us debug the issue. If we make these `Debug`, we'll lose these messages when logging at default `Info` level.

Reply:
@aaithal Theoretically...we shouldn't have async failures due to missing clients.  This discovery logic is used for capability registration, and if we don't register a capability there shouldn't be a task scheduled that requires an unavailable client.

Reply:
@samuelkarp yeah, but bugs.. Since this is just a once time discovery thing, logging it at Info shouldn't contribute to too much noise IMO.